packages/mdc-data-table/foundation.ts

```diff
@@ -0,0 +1,177 @@
+
+      this.setHeaderRowCheckboxState_();
+    }
+  }
+
+  /**
+   * @return Returns array of row elements.
+   */
+  getRows(): HTMLElement[] {
+    return this.adapter_.getRowElements();
+  }
+
+  /**
+   * Sets selected row ids. Overwrites previously selected rows.
+   * @param rowIds Array of row ids that needs to be selected.
+   */
+  setSelectedRowIds(rowIds: string[]) {
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      const rowId = this.adapter_.getRowIdAtIndex(rowIndex);
+
+      let isSelected = false;
+      if (rowId && rowIds.indexOf(rowId) >= 0) {
+        isSelected = true;
+      }
+
+      this.adapter_.setRowCheckboxCheckedAtIndex(rowIndex, isSelected);
+      this.selectRowAtIndex_(rowIndex, isSelected);
+    }
+
+    this.setHeaderRowCheckboxState_();
+  }
+
+  /**
+   * @return Returns array of selected row ids.
+   */
+  getSelectedRowIds(): Array<string|null> {
+    const selectedRowIds: Array<string|null> = [];
+    for (let rowIndex = 0; rowIndex < this.adapter_.getSelectedRowCount(); rowIndex++) {
+      selectedRowIds.push(this.adapter_.getRowIdAtIndex(rowIndex));
+    }
+
+    return selectedRowIds;
+  }
+
+  /**
+   * Handles header row checkbox change event.
+   */
+  handleHeaderRowCheckboxChange() {
+    const isHeaderChecked = this.adapter_.isHeaderRowCheckboxChecked();
+
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      this.adapter_.setRowCheckboxCheckedAtIndex(rowIndex, isHeaderChecked);
+      this.selectRowAtIndex_(rowIndex, isHeaderChecked);
+    }
+
+    if (isHeaderChecked) {
+      this.adapter_.notifySelectedAll();
+    } else {
+      this.adapter_.notifyUnselectedAll();
+    }
+  }
+
+  /**
+   * Handles change event originated from row checkboxes.
+   */
+  handleRowCheckboxChange(event: Event) {
+    const rowIndex = this.adapter_.getRowIndexByChildElement(event.target as HTMLInputElement);
+
+    if (rowIndex === -1) {
+      return;
+    }
+
+    const checkboxNativeEl = event.target as HTMLInputElement;
+    if (!checkboxNativeEl) {
+      return;
+    }
+
+    const selected = checkboxNativeEl.checked;
+
+    this.selectRowAtIndex_(rowIndex, selected);
+    this.setHeaderRowCheckboxState_();
+
+    const rowId = this.adapter_.getRowIdAtIndex(rowIndex);
+    this.adapter_.notifyRowSelectionChanged({rowId, rowIndex, selected});
+  }
+
+  /**
+   * Updates header row checkbox state based on number of rows selected.
+   */
+  private setHeaderRowCheckboxState_() {
+    if (this.adapter_.getSelectedRowCount() === this.adapter_.getRowCount()) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(true);
+    } else if (this.adapter_.getSelectedRowCount() === 0) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(false);
+    } else {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(true);
```

Comment:
Explicit call to `this.adapter_.setHeaderRowCheckboxChecked(false)` here, so we don't bake in the assumption that setting indeterminate to true will handle it.

Reply:
Done.